minikube update-context: invalid memory address or nil pointer dereference

`minikube update-context` command occurs the following error:
```
 panic: runtime error: invalid memory address or nil pointer dereference
 [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x523891f]
 
 goroutine 1 [running]:
 k8s.io/minikube/pkg/minikube/kubeconfig.UpdateEndpoint(0x57bad36, 0x8, 0xc000bd9410, 0xe, 0x20fb, 0xc00013f560, 0x19, 0x5b32f40, 0xc000bdd560, 0xc000bd9410)
         /private/tmp/minikube-20200812-94589-17snemo/pkg/minikube/kubeconfig/kubeconfig.go:143 +0x2af
 k8s.io/minikube/cmd/minikube/cmd.glob..func31(0x6927200, 0x696cf80, 0x0, 0x0)
         /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/cmd/update-context.go:37 +0x10a
 github.com/spf13/cobra.(*Command).execute(0x6927200, 0x696cf80, 0x0, 0x0, 0x6927200, 0x696cf80)
         /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x29d
 github.com/spf13/cobra.(*Command).ExecuteC(0x6925280, 0x0, 0x1, 0xc00024fbe0)
         /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
 github.com/spf13/cobra.(*Command).Execute(...)
         /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
 k8s.io/minikube/cmd/minikube/cmd.Execute()
         /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/cmd/root.go:106 +0x72c
 main.main()
         /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/main.go:71 +0x11f
```

<!--- Please include the "minikube start" command you used in your reproduction steps --->
**Steps to reproduce the issue:** 

```
$ minikube start --vm-driver virtualbox
😄  minikube v1.12.3 on Darwin 10.15.5
✨  Using the virtualbox driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🔥  Creating virtualbox VM (CPUs=2, Memory=6000MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.18.3 on Docker 19.03.12 ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube"
$ cat <<EOL > kubeconfig
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /home/la-croix/apiserver.crt
    server: 192.168.1.1:8080
  name: la-croix
contexts:
- context:
    cluster: la-croix
    user: la-croix
  name: la-croix
current-context: la-croix
kind: Config
preferences: {}
users:
- name: la-croix
  user:
    client-certificate: /home/la-croix/apiserver.crt
    client-key: /home/la-croix/apiserver.key
EOL
$ KUBECONFIG=./kubeconfig minikube update-context
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x523891f]

goroutine 1 [running]:
k8s.io/minikube/pkg/minikube/kubeconfig.UpdateEndpoint(0x57bad36, 0x8, 0xc000c4c3a0, 0xe, 0x20fb, 0xc00003e22b, 0xc, 0x5b32f40, 0xc0003d2000, 0xc000c4c3a0)
        /private/tmp/minikube-20200812-94589-17snemo/pkg/minikube/kubeconfig/kubeconfig.go:143 +0x2af
k8s.io/minikube/cmd/minikube/cmd.glob..func31(0x6927200, 0x696cf80, 0x0, 0x0)
        /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/cmd/update-context.go:37 +0x10a
github.com/spf13/cobra.(*Command).execute(0x6927200, 0x696cf80, 0x0, 0x0, 0x6927200, 0x696cf80)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0x6925280, 0x0, 0x1, 0xc00053a0c0)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
k8s.io/minikube/cmd/minikube/cmd.Execute()
        /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/cmd/root.go:106 +0x72c
main.main()
        /private/tmp/minikube-20200812-94589-17snemo/cmd/minikube/main.go:71 +0x11f
```

<!--- TIP: Add the "--alsologtostderr" flag to the command-line for more logs --->
**Full output of failed command:** 



**Full output of `minikube start` command used, if not already included:**



**Optional: Full output of `minikube logs` command:**
<details>


</details>

--- a/go.mod
+++ b/go.mod
@@ -72,7 +72,6 @@ require (
 	github.com/zchee/go-vmnet v0.0.0-20161021174912-97ebf9174097
 	golang.org/x/build v0.0.0-20190927031335-2835ba2e683f
 	golang.org/x/crypto v0.0.0-20200510223506-06a226fb4e37
-	golang.org/x/exp v0.0.0-20200224162631-6cc2880d07d6 // indirect
 	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
 	golang.org/x/sync v0.0.0-20200317015054-43a5402ce75a
 	golang.org/x/sys v0.0.0-20200523222454-059865788121

--- a/pkg/minikube/kubeconfig/kubeconfig.go
+++ b/pkg/minikube/kubeconfig/kubeconfig.go
@@ -21,7 +21,6 @@ import (
 	"io/ioutil"
 	"net/url"
 	"os"
-	"path"
 	"path/filepath"
 	"strconv"
 
@@ -31,7 +30,6 @@ import (
 	"k8s.io/client-go/tools/clientcmd/api"
 	"k8s.io/client-go/tools/clientcmd/api/latest"
 	"k8s.io/minikube/pkg/minikube/constants"
-	"k8s.io/minikube/pkg/minikube/localpath"
 	pkgutil "k8s.io/minikube/pkg/util"
 	"k8s.io/minikube/pkg/util/lock"
 )
@@ -121,26 +119,11 @@ func UpdateEndpoint(contextName string, hostname string, port int, confpath stri
 		return false, errors.Wrap(err, "read")
 	}
 
-	address := "https://" + hostname + ":" + strconv.Itoa(port)
-
-	// if the kubeconfig is missed, create new one
-	if len(cfg.Clusters) == 0 {
-		lp := localpath.Profile(contextName)
-		gp := localpath.MiniPath()
-		kcs := &Settings{
-			ClusterName:          contextName,
-			ClusterServerAddress: address,
-			ClientCertificate:    path.Join(lp, "client.crt"),
-			ClientKey:            path.Join(lp, "client.key"),
-			CertificateAuthority: path.Join(gp, "ca.crt"),
-			KeepContext:          false,
-		}
-		err = PopulateFromSettings(kcs, cfg)
-		if err != nil {
-			return false, errors.Wrap(err, "populating kubeconfig")
-		}
+	if _, ok := cfg.Clusters[contextName]; !ok {
+		return false, errors.Errorf("%q does not appear in %s", contextName, confpath)
 	}
-	cfg.Clusters[contextName].Server = address
+
+	cfg.Clusters[contextName].Server = "https://" + hostname + ":" + strconv.Itoa(port)
 	err = writeToFile(cfg, confpath)
 	if err != nil {
 		return false, errors.Wrap(err, "write")

--- a/pkg/minikube/kubeconfig/kubeconfig.go
+++ b/pkg/minikube/kubeconfig/kubeconfig.go
@@ -21,6 +21,7 @@ import (
 	"io/ioutil"
 	"net/url"
 	"os"
+	"path"
 	"path/filepath"
 	"strconv"
 
@@ -30,6 +31,7 @@ import (
 	"k8s.io/client-go/tools/clientcmd/api"
 	"k8s.io/client-go/tools/clientcmd/api/latest"
 	"k8s.io/minikube/pkg/minikube/constants"
+	"k8s.io/minikube/pkg/minikube/localpath"
 	pkgutil "k8s.io/minikube/pkg/util"
 	"k8s.io/minikube/pkg/util/lock"
 )
@@ -119,11 +121,28 @@ func UpdateEndpoint(contextName string, hostname string, port int, confpath stri
 		return false, errors.Wrap(err, "read")
 	}
 
+	address := "https://" + hostname + ":" + strconv.Itoa(port)
+
+	// if the cluster setting is missed in the kubeconfig, create new one
 	if _, ok := cfg.Clusters[contextName]; !ok {
-		return false, errors.Errorf("%q does not appear in %s", contextName, confpath)
+		lp := localpath.Profile(contextName)
+		gp := localpath.MiniPath()
+		kcs := &Settings{
+			ClusterName:          contextName,
+			ClusterServerAddress: address,
+			ClientCertificate:    path.Join(lp, "client.crt"),
+			ClientKey:            path.Join(lp, "client.key"),
+			CertificateAuthority: path.Join(gp, "ca.crt"),
+			KeepContext:          false,
+		}
+		err = PopulateFromSettings(kcs, cfg)
+		if err != nil {
+			return false, errors.Wrap(err, "populating kubeconfig")
+		}
+	} else {
+		cfg.Clusters[contextName].Server = address
 	}
 
-	cfg.Clusters[contextName].Server = "https://" + hostname + ":" + strconv.Itoa(port)
 	err = writeToFile(cfg, confpath)
 	if err != nil {
 		return false, errors.Wrap(err, "write")

--- a/pkg/minikube/kubeconfig/kubeconfig_test.go
+++ b/pkg/minikube/kubeconfig/kubeconfig_test.go
@@ -343,6 +343,18 @@ func TestUpdateIP(t *testing.T) {
 			if test.status != statusActual {
 				t.Errorf("Expected status %t, but got %t", test.status, statusActual)
 			}
+
+			actual, err := readOrNew(configFilename)
+			if err != nil {
+				t.Fatal(err)
+			}
+			expected, err := decode(test.expCfg)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !configEquals(actual, expected) {
+				t.Errorf("Expected cfg %v, but got %v", expected, actual)
+			}
 		})
 
 	}

--- a/pkg/minikube/kubeconfig/kubeconfig_test.go
+++ b/pkg/minikube/kubeconfig/kubeconfig_test.go
@@ -25,6 +25,7 @@ import (
 
 	"k8s.io/client-go/tools/clientcmd/api"
 	"k8s.io/minikube/pkg/minikube/constants"
+	"k8s.io/minikube/pkg/minikube/localpath"
 
 	"k8s.io/client-go/tools/clientcmd"
 )
@@ -49,6 +50,40 @@ users:
   user:
     client-certificate: /home/la-croix/apiserver.crt
     client-key: /home/la-croix/apiserver.key
+`)
+
+var kubeConfigWithoutHTTPSUpdated = []byte(`
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority: /home/la-croix/apiserver.crt
+    server: 192.168.1.1:8080
+  name: la-croix
+- cluster:
+    certificate-authority: /home/la-croix/.minikube/ca.crt
+    server: https://192.168.10.100:8080
+  name: minikube
+contexts:
+- context:
+    cluster: la-croix
+    user: la-croix
+  name: la-croix
+- context:
+    cluster: minikube
+    user: minikube
+  name: minikube
+current-context: minikube
+kind: Config
+preferences: {}
+users:
+- name: la-croix
+  user:
+    client-certificate: /home/la-croix/apiserver.crt
+    client-key: /home/la-croix/apiserver.key
+- name: minikube
+  user:
+    client-certificate: /home/la-croix/.minikube/profiles/minikube/client.crt
+    client-key: /home/la-croix/.minikube/profiles/minikube/client.key
 `)
 
 var kubeConfig192 = []byte(`
@@ -115,6 +150,37 @@ users:
   user:
     client-certificate: /home/la-croix/apiserver.crt
     client-key: /home/la-croix/apiserver.key
+`)
+
+var kubeConfigNoClusters = []byte(`
+apiVersion: v1
+clusters:
+contexts:
+kind: Config
+preferences: {}
+users:
+`)
+
+var kubeConfigNoClustersUpdated = []byte(`
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority: /home/la-croix/.minikube/ca.crt
+    server: https://192.168.10.100:8080
+  name: minikube
+contexts:
+- context:
+    cluster: minikube
+    user: minikube
+  name: minikube
+current-context: minikube
+kind: Config
+preferences: {}
+users:
+- name: minikube
+  user:
+    client-certificate: /home/la-croix/.minikube/profiles/minikube/client.crt
+    client-key: /home/la-croix/.minikube/profiles/minikube/client.key
 `)
 
 func TestUpdate(t *testing.T) {
@@ -300,8 +366,8 @@ func TestUpdateIP(t *testing.T) {
 			hostname:    "192.168.10.100",
 			port:        8080,
 			existing:    kubeConfigWithoutHTTPS,
-			err:         true,
-			expCfg:      kubeConfigWithoutHTTPS,
+			status:      true,
+			expCfg:      kubeConfigWithoutHTTPSUpdated,
 		},
 		{
 			description: "same IP",
@@ -326,7 +392,18 @@ func TestUpdateIP(t *testing.T) {
 			status:      true,
 			expCfg:      kubeConfigLocalhost12345,
 		},
+		{
+			description: "no clusters",
+			hostname:    "192.168.10.100",
+			port:        8080,
+			existing:    kubeConfigNoClusters,
+			status:      true,
+			expCfg:      kubeConfigNoClustersUpdated,
+		},
 	}
+
+	os.Setenv(localpath.MinikubeHome, "/home/la-croix")
+
 	for _, test := range tests {
 		test := test
 		t.Run(test.description, func(t *testing.T) {
@@ -353,7 +430,7 @@ func TestUpdateIP(t *testing.T) {
 				t.Fatal(err)
 			}
 			if !configEquals(actual, expected) {
-				t.Errorf("Expected cfg %v, but got %v", expected, actual)
+				t.Fatal("configs did not match")
 			}
 		})
 

--- a/pkg/minikube/kubeconfig/kubeconfig_test.go
+++ b/pkg/minikube/kubeconfig/kubeconfig_test.go
@@ -328,6 +328,7 @@ func TestUpdateIP(t *testing.T) {
 		},
 	}
 	for _, test := range tests {
+		test := test
 		t.Run(test.description, func(t *testing.T) {
 			t.Parallel()
 			configFilename := tempFile(t, test.existing)

--- a/test/integration/functional_test.go
+++ b/test/integration/functional_test.go
@@ -1007,14 +1007,86 @@ func validateCertSync(ctx context.Context, t *testing.T, profile string) {
 func validateUpdateContextCmd(ctx context.Context, t *testing.T, profile string) {
 	defer PostMortemLogs(t, profile)
 
-	rr, err := Run(t, exec.CommandContext(ctx, Target(), "-p", profile, "update-context", "--alsologtostderr", "-v=2"))
-	if err != nil {
-		t.Errorf("failed to run minikube update-context: args %q: %v", rr.Command(), err)
+	tests := []struct {
+		name       string
+		kubeconfig []byte
+		want       []byte
+	}{
+		{
+			name:       "no changes",
+			kubeconfig: nil,
+			want:       []byte("No changes"),
+		},
+		{
+			name: "no minikube cluster",
+			kubeconfig: []byte(`
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority: /home/la-croix/apiserver.crt
+    server: 192.168.1.1:8080
+  name: la-croix
+contexts:
+- context:
+    cluster: la-croix
+    user: la-croix
+  name: la-croix
+current-context: la-croix
+kind: Config
+preferences: {}
+users:
+- name: la-croix
+  user:
+    client-certificate: /home/la-croix/apiserver.crt
+    client-key: /home/la-croix/apiserver.key
+`),
+			want: []byte("context has been updated"),
+		},
+		{
+			name: "no clusters",
+			kubeconfig: []byte(`
+apiVersion: v1
+clusters:
+contexts:
+kind: Config
+preferences: {}
+users:
+`),
+			want: []byte("context has been updated"),
+		},
 	}
 
-	want := []byte("No changes")
-	if !bytes.Contains(rr.Stdout.Bytes(), want) {
-		t.Errorf("update-context: got=%q, want=*%q*", rr.Stdout.Bytes(), want)
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			c := exec.CommandContext(ctx, Target(), "-p", profile, "update-context", "--alsologtostderr", "-v=2")
+			if tc.kubeconfig != nil {
+				tf, err := ioutil.TempFile("", "kubeconfig")
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if err := ioutil.WriteFile(tf.Name(), tc.kubeconfig, 0644); err != nil {
+					t.Fatal(err)
+				}
+
+				t.Cleanup(func() {
+					os.Remove(tf.Name())
+				})
+
+				c.Env = append(os.Environ(), fmt.Sprintf("KUBECONFIG=%s", tf.Name()))
+			}
+
+			rr, err := Run(t, c)
+			if err != nil {
+				t.Errorf("failed to run minikube update-context: args %q: %v", rr.Command(), err)
+			}
+
+			if !bytes.Contains(rr.Stdout.Bytes(), tc.want) {
+				t.Errorf("update-context: got=%q, want=*%q*", rr.Stdout.Bytes(), tc.want)
+			}
+		})
 	}
 }
 

